<!doctype html>
<html class="no-js" lang="en">

<head>
  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-134228602-6"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'UA-134228602-6');
  </script>

  <meta charset="utf-8">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <title>üßòüèª üîß üö£üèª [DotNetBook] Excepciones: Arquitectura de sistema de tipo üë®üèΩ‚Äçüé® üë®üèø‚Äçüé® üßëüèº‚Äçü§ù‚Äçüßëüèº</title>
  <link rel="icon" type="image/x-icon" href="/favicon.ico" />
  <meta name="description" content="Con este art√≠culo, contin√∫o publicando una serie de art√≠culos, cuyo resultado ser√° un libro sobre el trabajo de .NET CLR y .NET en general. Para enlac...">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <link rel="stylesheet" href="../../css/main.css">

  <link href="https://fonts.googleapis.com/css?family=Quicksand&display=swap" rel="stylesheet">

  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <script>window.jQuery || document.write('<script src="../../js/vendors/jquery-3.3.1.min.js"><\/script>')</script>

  <script>document.write('<script src="//pagea' + 'd2.googles' + 'yndication.com/pagea' + 'd/js/a' + 'dsby' + 'google.js"><\/script>')</script>
  <script>
        var superSpecialObject = {};
        superSpecialObject['google_a' + 'd_client'] = 'ca-p' + 'ub-6974184241884155';
        superSpecialObject['enable_page_level_a' + 'ds'] = true;
       (window['a' + 'dsbygoogle'] = window['a' + 'dsbygoogle'] || []).push(superSpecialObject);
  </script>
</head>

<body>
  <!--[if lte IE 9]>
    <p class="browserupgrade">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience and security.</p>
  <![endif]-->
  <header class="page-header js-page-header">
    <a class="page-header-logo-container" href="https://weekly-geekly-es.github.io/index.html"></a>
    <div class="page-header-text">Geekly articles weekly</div>
  </header>
  <section class="page js-page"><h1>[DotNetBook] Excepciones: Arquitectura de sistema de tipo</h1><div class="post__body post__body_full"><div class="post__text post__text-html post__text_v1" id="post-content-body" data-io-article-url="https://habr.com/ru/company/clrium/blog/419927/"><p><img width="350" src="https://habrastorage.org/webt/vh/7j/tq/vh7jtqhzbne4h3rjhprca2pruhu.png" align="left">  Con este art√≠culo, contin√∫o publicando una serie de art√≠culos, cuyo resultado ser√° un libro sobre el trabajo de .NET CLR y .NET en general.  Para enlaces, bienvenidos a cat. </p><br><h2 id="arhitektura-isklyuchitelnoy-situacii">  Arquitectura de excepci√≥n </h2><br><p> Probablemente uno de los temas m√°s importantes con respecto al tema de las excepciones es el tema de construir una arquitectura de excepci√≥n en su aplicaci√≥n.  Esta pregunta es interesante por muchas razones.  En cuanto a m√≠, lo principal es la aparente simplicidad con la que no siempre es obvio qu√© hacer.  Esta propiedad es inherente a todas las construcciones b√°sicas que se usan en todas partes: es <code>IEnumerable</code> e <code>IDisposable</code> and <code>IObservable</code> y otras.  Por un lado, est√°n haciendo se√±as por su simplicidad, involucr√°ndose en su uso en una variedad de situaciones.  Y, por otro lado, est√°n llenos de remolinos y vados, de los cuales, sin saber c√≥mo, a veces no salen.  Y, tal vez, mirando el volumen futuro, su pregunta ha madurado: entonces, ¬øqu√© es en situaciones excepcionales? </p><br><blockquote><h3>  Nota </h3><br>  El cap√≠tulo publicado en Habr√© no est√° actualizado y, probablemente, est√° un poco desactualizado.  Y, por lo tanto, consulte el original para obtener un texto m√°s reciente: <br><br><ul><li><img src="https://habrastorage.org/webt/3q/6g/qa/3q6gqaz40qx-jzscjf3jbxatxhg.png">  Libro CLR: <a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=es&amp;u=">GitHub, tabla de contenido</a> </li><li><img src="https://habrastorage.org/webt/3q/6g/qa/3q6gqaz40qx-jzscjf3jbxatxhg.png">  Libro CLR: <a href="">GitHub, cap√≠tulo</a> </li><li><img src="https://habrastorage.org/webt/eo/6g/eo/eo6geog0tg5ernqmv2lcmufefta.png">  Libros de la versi√≥n 0.5.2, PDF: <a href="">versi√≥n de GitHub</a> </li></ul><br></blockquote><p> <a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=es&amp;u="><img src="https://habrastorage.org/getpro/habr/post_images/97f/1d3/cf0/97f1d3cf0e2a6bf007066eb60a789c31.png"></a> </p><a name="habracut"></a><br><p>  Pero para llegar a algunas conclusiones con respecto a la construcci√≥n de la arquitectura de clases de situaciones excepcionales, debemos acumular cierta experiencia con respecto a su clasificaci√≥n.  Despu√©s de todo, solo despu√©s de haber entendido con qu√© trataremos, c√≥mo y en qu√© situaciones el programador debe elegir el tipo de error, y en qu√©: hacer la elecci√≥n con respecto a la captura u omisi√≥n de excepciones, puede comprender c√≥mo puede construir un sistema de tipos de tal manera que sea obvio para su c√≥digo  Por lo tanto, trataremos de clasificar situaciones excepcionales (no los tipos de excepciones en s√≠, sino precisamente las situaciones) de acuerdo con varios criterios. </p><br><h3 id="po-teoreticheskoy-vozmozhnosti-perehvata-proektiruemogo-isklyucheniya">  Seg√∫n la posibilidad te√≥rica de atrapar la excepci√≥n proyectada </h3><br><p>  En t√©rminos de intercepci√≥n te√≥rica, las excepciones se pueden dividir f√°cilmente en dos tipos: las que interceptar√°n con precisi√≥n y las que tienen mayor probabilidad de interceptar.  ¬øPor qu√© <em>con un alto grado de probabilidad</em> ?  Porque siempre habr√° alguien que intente interceptar, aunque esto no tuvo que hacerse por completo. </p><br><p>  Revelemos primero las caracter√≠sticas del primer grupo: excepciones que deber√≠an capturar. </p><br><p>  Cuando introducimos una excepci√≥n de este tipo, por un lado informamos al subsistema externo que estamos en una posici√≥n en la que otras acciones dentro de nuestros datos no tienen sentido.  Por otro lado, queremos decir que no se rompi√≥ nada global y si somos eliminados, nada cambiar√° y, por lo tanto, esta excepci√≥n se puede interceptar f√°cilmente para mejorar la situaci√≥n.  Esta propiedad es muy importante: determina la importancia del error y la creencia de que si detecta la excepci√≥n y simplemente borra los recursos, puede ejecutar el c√≥digo de manera segura. </p><br><p>  El segundo grupo, no importa cu√°n extra√±o pueda sonar, es responsable de las excepciones que no necesitan ser atrapadas.  Solo se pueden usar para escribir en el registro de errores, pero no para corregir de alguna manera la situaci√≥n.  El ejemplo m√°s simple son las excepciones de grupo <code>ArgumentException</code> y <code>NullReferenceException</code> .  De hecho, en una situaci√≥n normal, no debe, por ejemplo, atrapar la excepci√≥n <code>ArgumentNullException</code> porque la fuente del problema aqu√≠ ser√° usted, y nadie m√°s.  Si detecta esta excepci√≥n, entonces asume que cometi√≥ un error y le dio el m√©todo que no pudo darle a: </p><br><pre> <code class="plaintext hljs">void SomeMethod(object argument) { try { AnotherMethod(argument); } catch (ArgumentNullException exception) { // Log it } }</code> </pre> <br><p>  En este m√©todo, intentamos atrapar una <code>ArgumentNullException</code> .  Pero en mi opini√≥n, su intercepci√≥n parece muy extra√±a: lanzar los argumentos correctos al m√©todo es nuestra preocupaci√≥n.  No ser√≠a correcto reaccionar despu√©s del hecho: en tal situaci√≥n, lo m√°s correcto que se puede hacer es verificar los datos transmitidos por adelantado, antes de llamar al m√©todo, o incluso mejor, para construir el c√≥digo de tal manera que simplemente no sea posible recibir par√°metros incorrectos. </p><br><p>  Otro grupo es la eliminaci√≥n de errores fatales.  Si un determinado cach√© est√° roto y el funcionamiento del subsistema no ser√° correcto en ning√∫n caso?  Entonces este es un error fatal y no se garantizar√° que el c√≥digo m√°s cercano a la pila lo intercepte: </p><br><pre> <code class="plaintext hljs">T GetFromCacheOrCalculate() { try { if(_cache.TryGetValue(Key, out var result)) { return result; } else { T res = Strategy(Key); _cache[Key] = res; return res; } } catch (CacheCorreptedException exception) { RecreateCache(); return GetFromCacheOrCalculate(); } }</code> </pre> <br><p>  Y deje que <code>CacheCorreptedException</code> sea ‚Äã‚Äãuna excepci√≥n que significa "el cach√© en el disco duro no es consistente".  Luego resulta que si la causa de tal error es fatal para el subsistema de almacenamiento en cach√© (por ejemplo, no hay permisos para el archivo de cach√©), entonces el c√≥digo adicional si no puede recrear el cach√© con el comando <code>RecreateCache</code> , y por lo tanto el hecho de atrapar esta excepci√≥n es un error en s√≠ mismo. </p><br><h3 id="po-fakticheskomu-perehvatu-isklyuchitelnoy-situacii">  Sobre la intercepci√≥n real de una excepci√≥n </h3><br><p>  Otra pregunta que detiene nuestro vuelo de pensamiento en los algoritmos de programaci√≥n es la comprensi√≥n: ¬øvale la pena atrapar estas u otras excepciones o vale la pena alguien que entiende dejarlos pasar?  Traduciendo al lenguaje de los t√©rminos, la pregunta que debemos resolver es distinguir entre √°reas de responsabilidad.  Veamos el siguiente c√≥digo: </p><br><pre> <code class="plaintext hljs">namespace JetFinance.Strategies { public class WildStrategy : StrategyBase { private Random random = new Random(); public void PlayRussianRoulette() { if(DateTime.Now.Second == (random.Next() % 60)) { throw new StrategyException(); } } } public class StrategyException : Exception { /* .. */ } } namespace JetFinance.Investments { public class WildInvestment { WildStrategy _strategy; public WildInvestment(WildStrategy strategy) { _strategy = strategy; } public void DoSomethingWild() { ?try? { _strategy.PlayRussianRoulette(); } catch(StrategyException exception) { } } } } using JetFinance.Strategies; using JetFinance.Investments; void Main() { var foo = new WildStrategy(); var boo = new WildInvestment(foo); ?try? { boo.DoSomethingWild(); } catch(StrategyException exception) { } }</code> </pre><br><p>  ¬øCu√°l de las dos estrategias propuestas es m√°s correcta?  El √°rea de responsabilidad es muy importante.  Inicialmente, puede parecer que, dado que el trabajo de <code>WildInvestment</code> y su consistencia dependen completamente de <code>WildStrategy</code> , si <code>WildInvestment</code> simplemente ignora esta excepci√≥n, pasar√° a un nivel superior y no hay necesidad de hacer nada m√°s.  Sin embargo, tenga en cuenta que existe un problema puramente arquitect√≥nico: el m√©todo <code>Main</code> detecta una excepci√≥n de una capa arquitect√≥nicamente invocando un m√©todo arquitect√≥nicamente diferente.  ¬øC√≥mo se ve en t√©rminos de uso?  S√≠, en general, se ve as√≠: </p><br><ul><li>  la preocupaci√≥n por esta excepci√≥n fue simplemente superada por nosotros; </li><li>  el usuario de esta clase no est√° seguro de que esta excepci√≥n se arroje a trav√©s de varios m√©todos que tenemos ante nosotros espec√≠ficamente </li><li>  comenzamos a dibujar adicciones innecesarias, de las cuales nos deshacemos, causando una capa intermedia. </li></ul><br><p>  Sin embargo, otra conclusi√≥n se deduce de esta conclusi√≥n: debemos establecer <code>catch</code> en el m√©todo <code>DoSomethingWild</code> .  Y esto es algo extra√±o para nosotros: <code>WildInvestment</code> parece depender mucho de alguien.  Es decir  si <code>PlayRussianRoulette</code> no pudo funcionar, entonces <code>DoSomethingWild</code> tambi√©n: no tiene c√≥digos de retorno, pero debe jugar a la ruleta.  ¬øQu√© hacer en una situaci√≥n tan desesperada?  La respuesta es realmente simple: al estar en otra capa, <code>DoSomethingWild</code> deber√≠a lanzar su propia excepci√≥n, que se refiere a esta capa y envolver el original como la fuente original del problema, en <code>InnerException</code> : </p><br><pre> <code class="plaintext hljs">namespace JetFinance.Strategies { pubilc class WildStrategy { private Random random = new Random(); public void PlayRussianRoulette() { if(DateTime.Now.Second == (random.Next() % 60)) { throw new StrategyException(); } } } public class StrategyException : Exception { /* .. */ } } namespace JetFinance.Investments { public class WildInvestment { WildStrategy _strategy; public WildInvestment(WildStrategy strategy) { _strategy = strategy; } public void DoSomethingWild() { try { _strategy.PlayRussianRoulette(); } catch(StrategyException exception) { throw new FailedInvestmentException("Oops", exception); } } } public class InvestmentException : Exception { /* .. */ } public class FailedInvestmentException : Exception { /* .. */ } } using JetFinance.Investments; void Main() { var foo = new WildStrategy(); var boo = new WildInvestment(foo); try { boo.DoSomethingWild(); } catch(FailedInvestmentException exception) { } }</code> </pre> <br><p>  Volviendo la excepci√≥n a otra, esencialmente transferimos los problemas de una capa de aplicaci√≥n a otra, haciendo que su trabajo sea m√°s predecible desde el punto de vista del usuario de esta clase: el m√©todo <code>Main</code> . </p><br><h3 id="po-voprosam-pereispolzovaniya">  Por problemas de reutilizaci√≥n </h3><br><p>  Muy a menudo nos enfrentamos a una tarea dif√≠cil: por un lado, somos demasiado flojos para crear un nuevo tipo de excepci√≥n, y cuando decidimos, no siempre est√° claro de qu√© empujar: qu√© tipo tomar como base como base.  Pero son precisamente estas decisiones las que determinan toda la arquitectura de situaciones excepcionales.  Repasemos las soluciones populares y saquemos algunas conclusiones. </p><br><p>  Al elegir el tipo de excepciones, puede intentar tomar una soluci√≥n ya existente: encuentre una excepci√≥n con un significado similar en el nombre y √∫sela.  Por ejemplo, si nos dieron una entidad a trav√©s de un par√°metro que de alguna manera no nos conviene, podemos lanzar una <code>InvalidArgumentException</code> , que indica la causa del error en el Mensaje.  Este escenario se ve bien, especialmente teniendo en cuenta que <code>InvalidArgumentException</code> est√° en el grupo de excepciones que no est√°n sujetas a capturas obligatorias.  Pero elegir <code>InvalidDataException</code> ser√° malo si est√° trabajando con alg√∫n dato.  Solo porque este tipo est√° en la zona <code>System.IO</code> , y esto no es lo que haces.  Es decir  Resulta que encontrar el tipo existente porque perezosamente hacer el tuyo casi siempre ser√° el enfoque equivocado.  Casi no hay excepciones que se crean para el c√≠rculo general de tareas.  Casi todos est√°n creados para situaciones espec√≠ficas y su reutilizaci√≥n ser√° una violaci√≥n grave de la arquitectura de situaciones excepcionales.  No solo eso, despu√©s de haber recibido una excepci√≥n de cierto tipo (por ejemplo, el mismo <code>System.IO.InvalidDataException</code> ), el usuario estar√° confundido: por un lado, ver√° el origen del problema en <code>System.IO</code> como un espacio de nombres de excepci√≥n y, por otro, un espacio de nombres de punto de lanzamiento completamente diferente.  Adem√°s, pensando en las reglas para lanzar esta excepci√≥n, ir√° a <a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=es&amp;u=">referencesource.microsoft.com</a> y encontrar√° <a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=es&amp;u=">todos los lugares donde se lanza</a> : </p><br><ul><li> <code>internal class System.IO.Compression.Inflater</code> </li> </ul><br><p>  Y √©l entender√° que <del>  solo alguien tiene las manos torcidas </del>  la elecci√≥n del tipo de excepci√≥n lo confundi√≥, ya que el m√©todo que arroj√≥ la excepci√≥n no estaba involucrado en la compresi√≥n. </p><br><p>  Adem√°s, para simplificar la reutilizaci√≥n, simplemente puede tomar y crear una √∫nica excepci√≥n declarando un campo <code>ErrorCode</code> con un c√≥digo de error y vivir feliz para siempre.  Parecer√≠a: una buena soluci√≥n.  Lanza la misma excepci√≥n en todas partes, configura el c√≥digo, captura solo una <code>catch</code> lo que aumenta la estabilidad de la aplicaci√≥n: y no hay nada m√°s que hacer.  Sin embargo, no est√© de acuerdo con esta posici√≥n.  Actuando de esta manera a lo largo de la aplicaci√≥n, por un lado, por supuesto, simplificas tu vida.  Pero, por otro lado, descarta la capacidad de capturar un subgrupo de excepciones, unidas por alguna caracter√≠stica com√∫n.  C√≥mo se hace esto, por ejemplo, con <code>ArgumentException</code> , que combina un grupo completo de excepciones por herencia.  El segundo inconveniente grave son las hojas de c√≥digo demasiado grandes e ilegibles que organizar√°n el filtrado por c√≥digo de error.  Pero si toma una situaci√≥n diferente: cuando la finalizaci√≥n del error no debe ser importante para el usuario final, la introducci√≥n de un tipo de generalizaci√≥n m√°s un c√≥digo de error parece una aplicaci√≥n mucho m√°s correcta: </p><br><pre> <code class="plaintext hljs">public class ParserException : Exception { public ParserError ErrorCode { get; } public ParserException(ParserError errorCode) { ErrorCode = errorCode; } public override string Message { get { return Resources.GetResource($"{nameof(ParserException)}{Enum.GetName(typeof(ParserError), ErrorCode)}"); } } } public enum ParserError { MissingModifier, MissingBracket, // ... } // Usage throw new ParserException(ParserError.MissingModifier);</code> </pre> <br><p>  El c√≥digo que protege la llamada del analizador casi siempre es indiferente por qu√© motivo se bloque√≥ el an√°lisis: el hecho del error en s√≠ mismo es importante para √©l.  Sin embargo, si esto se vuelve importante, el usuario siempre podr√° extraer el c√≥digo de error de la <code>ErrorCode</code> .  Para hacer esto, no es necesario buscar las palabras necesarias subcadenando en <code>Message</code> . </p><br><p>  Si comienza a ignorar los problemas de reutilizaci√≥n, puede crear un tipo de excepci√≥n para cada situaci√≥n.  Por un lado, parece l√≥gico: un tipo de error es un tipo de excepci√≥n.  Sin embargo, aqu√≠, como en todo, lo principal es no exagerar: tener operaciones excepcionales en cada punto de liberaci√≥n, causa problemas de intercepci√≥n: el c√≥digo del m√©todo de llamada se sobrecargar√° con bloques <code>catch</code> .  Despu√©s de todo, √©l necesita manejar todo tipo de excepciones que quieras darle.  Otra desventaja es puramente arquitect√≥nica.  Si no usa la herencia, desorienta al usuario de estas excepciones: puede haber mucho en com√∫n entre ellas y debe interceptarlas individualmente. </p><br><p>  Sin embargo, hay buenos escenarios para introducir tipos particulares para situaciones espec√≠ficas.  Por ejemplo, cuando ocurre un desglose no para toda la entidad como un todo, sino para un m√©todo espec√≠fico.  Entonces este tipo deber√≠a estar en la jerarqu√≠a de herencia en un lugar tal que no se piense en interceptarlo junto con otra cosa: por ejemplo, seleccion√°ndolo a trav√©s de una rama de herencia separada. </p><br><p>  Adem√°s, si combina estos dos enfoques, puede obtener una caja de herramientas muy poderosa para trabajar con un grupo de errores: puede introducir un tipo abstracto generalizado del cual heredar situaciones particulares espec√≠ficas.  La clase base (nuestro tipo de generalizaci√≥n) debe estar equipada con una propiedad abstracta que almacene el c√≥digo de error, y los herederos anular√°n esta propiedad para especificar este c√≥digo de error: </p><br><pre> <code class="plaintext hljs">public abstract class ParserException : Exception { public abstract ParserError ErrorCode { get; } public override string Message { get { return Resources.GetResource($"{nameof(ParserException)}{Enum.GetName(typeof(ParserError), ErrorCode)}"); } } } public enum ParserError { MissingModifier, MissingBracket } public class MissingModifierParserException : ParserException { public override ParserError ErrorCode { get; } =&gt; ParserError.MissingModifier; } public class MissingBracketParserException : ParserException { public override ParserError ErrorCode { get; } =&gt; ParserError.MissingBracket; } // Usage throw new MissingModifierParserException(ParserError.MissingModifier);</code> </pre> <br><p>  ¬øCu√°les son las maravillosas propiedades que obtenemos con este enfoque? </p><br><ul><li>  por un lado, conservamos la captura de una excepci√≥n por el tipo b√°sico; </li><li>  por otro lado, al detectar una excepci√≥n por el tipo b√°sico, todav√≠a era posible descubrir una situaci√≥n espec√≠fica; </li><li>  Adem√°s de todo, es posible interceptar un tipo espec√≠fico, y no uno b√°sico, sin utilizar la estructura plana de las clases. </li></ul><br><p>  En cuanto a m√≠, esta es una opci√≥n muy conveniente. </p><br><h3 id="po-otnosheniyu-k-edinoy-gruppe-povedencheskih-situaciy">  En relaci√≥n a un solo grupo de situaciones de comportamiento </h3><br><p>  ¬øQu√© conclusiones pueden extraerse en base al razonamiento descrito anteriormente?  Tratemos de formularlos: </p><br><p>  Para comenzar, decidamos qu√© se entiende por situaciones.  Cuando hablamos de clases y objetos, estamos acostumbrados principalmente a operar entidades con alg√∫n estado interno sobre el cual podemos llevar a cabo acciones.  Resulta que al hacerlo encontramos el primer tipo de situaci√≥n de comportamiento: acciones en una determinada entidad.  Adem√°s, si observa el gr√°fico de objetos como si fuera desde el exterior, notar√° que est√° l√≥gicamente combinado en grupos funcionales: el primero trata con el almacenamiento en cach√©, el segundo trata con bases de datos, el tercero realiza c√°lculos matem√°ticos.  Las capas pueden pasar por todos estos grupos funcionales: una capa de registro de varios estados internos, registro de procesos, rastreo de llamadas a m√©todos.  Las capas pueden ser m√°s amplias: combinar varios grupos funcionales.  Por ejemplo, una capa de modelo, una capa de controlador, una capa de presentaci√≥n.  Estos grupos pueden estar en el mismo ensamblaje, o en grupos completamente diferentes, pero cada uno de ellos puede crear sus propias situaciones excepcionales. </p><br><p>  Resulta que si argumenta de esta manera, puede construir una jerarqu√≠a de tipos de situaciones excepcionales, en funci√≥n del tipo que pertenece a un grupo o capa en particular, creando as√≠ la capacidad de capturar excepciones al c√≥digo para facilitar la navegaci√≥n sem√°ntica en esta jerarqu√≠a de tipos. </p><br><p>  Miremos el c√≥digo: </p><br><pre> <code class="plaintext hljs">namespace JetFinance { namespace FinancialPipe { namespace Services { namespace XmlParserService { } namespace JsonCompilerService { } namespace TransactionalPostman { } } } namespace Accounting { /* ... */ } }</code> </pre><br><p>  ¬øC√≥mo se ve?  En cuanto a m√≠, los espacios de nombres son una gran oportunidad para agrupar naturalmente tipos de excepciones de acuerdo con sus situaciones de comportamiento: todo lo que pertenece a ciertos grupos deber√≠a estar all√≠, incluidas las excepciones.  Adem√°s, cuando reciba una cierta excepci√≥n, adem√°s del nombre de su tipo, ver√° su espacio de nombres, que determinar√° claramente su afiliaci√≥n.  ¬øRecuerda el mal ejemplo de reutilizaci√≥n del tipo <code>InvalidDataException</code> que se define realmente en el espacio de nombres <code>System.IO</code> ?  Su pertenencia a este espacio de nombres significa que, en esencia, una excepci√≥n de este tipo puede ser eliminada de las clases ubicadas en el espacio de nombres <code>System.IO</code> o en una m√°s anidada.  Pero la excepci√≥n misma fue expulsada de un lugar completamente diferente, confundiendo al investigador del problema que surgi√≥.  Al enfocar los tipos de excepci√≥n en los mismos espacios de nombres que los tipos que arrojan estas excepciones, mantiene la arquitectura de tipo consistente por un lado y, por otro lado, facilita que el desarrollador final comprenda las razones de lo que sucedi√≥. </p><br><p>  ¬øCu√°l es la segunda forma de agrupar a nivel de c√≥digo?  Herencia: </p><br><pre> <code class="plaintext hljs">public abstract class LoggerExceptionBase : Exception { protected LoggerExceptionBase(..); } public class IOLoggerException : LoggerExceptionBase { internal IOLoggerException(..); } public class ConfigLoggerException : LoggerExceptionBase { internal ConfigLoggerException(..); }</code> </pre><br><p>  Adem√°s, si en el caso de las entidades de aplicaci√≥n ordinarias, herencia significa la herencia de comportamiento y datos, combinando los tipos al pertenecer a un <em>solo grupo de entidades</em> , entonces, en el caso de excepciones, herencia significa pertenecer a un <em>solo grupo de situaciones</em> , ya que la esencia de la excepci√≥n no es la esencia, sino la problem√°tica. </p><br><p>  Combinando ambos m√©todos de agrupaci√≥n, podemos sacar algunas conclusiones: </p><br><ul><li>  dentro del ensamblaje ( <code>Assembly</code> ) debe estar presente el tipo b√°sico de excepciones que arroja este ensamblaje.  Este tipo de excepci√≥n debe estar en el espacio de nombres ra√≠z para el ensamblado.  Esta ser√° la primera capa de la agrupaci√≥n; </li><li>  M√°s adentro del ensamblaje, puede haber uno o m√°s espacios de nombres diferentes.  Cada uno de ellos divide el ensamblaje en algunas zonas funcionales, definiendo as√≠ los grupos de situaciones que surgen en este ensamblaje.  Estas pueden ser zonas de controladores, entidades de bases de datos, algoritmos de procesamiento de datos y otros.  Para nosotros, estos espacios de nombres son una agrupaci√≥n de tipos por afiliaci√≥n funcional y, desde el punto de vista de las excepciones, una agrupaci√≥n por zonas problem√°ticas del mismo ensamblado; </li><li>  La herencia de excepciones puede ir solo de tipos en el mismo espacio de nombres o en la ra√≠z m√°s.  Esto garantiza una comprensi√≥n inequ√≠voca de la situaci√≥n por parte del usuario final y la ausencia de interceptaci√≥n de excepciones <em>dejadas</em> al interceptar seg√∫n el tipo b√°sico.<font style="vertical-align: inherit;"><font style="vertical-align: inherit;">De acuerdo: ser√≠a extra√±o </font></font><code>global::Finiki.Logistics.OhMyException</code><font style="vertical-align: inherit;"><font style="vertical-align: inherit;">tenerlo </font></font><code>catch(global::Legacy.LoggerExeption exception)</code><font style="vertical-align: inherit;"><font style="vertical-align: inherit;">, pero el siguiente c√≥digo se ve absolutamente armonioso:</font></font></li></ul><br><pre> <code class="plaintext hljs">namespace JetFinance.FinancialPipe { namespace Services.XmlParserService { public class XmlParserServiceException : FinancialPipeExceptionBase { // .. } public class Parser { public void Parse(string input) { // .. } } } public abstract class FinancialPipeExceptionBase : Exception { } } using JetFinance.FinancialPipe; using JetFinance.FinancialPipe.Services.XmlParserService; var parser = new Parser(); try { parser.Parse(); } catch (XmlParserServiceException exception) { // Something wrong in parser } catch (FinancialPipeExceptionBase exception) { // Something else wrong. Looks critical because we don't know real reason }</code> </pre><br><p> ,   :        , ,   ,       <code>XmlParserServiceException</code> . ,   ,      ,  <code>JetFinance.FinancialPipe.FinancialPipeExceptionBase</code> ,       :    <code>XmlParserService</code>    ,       .          ,               <code>catch</code>      :          . </p><br><p>     ? </p><br><ul><li>        .     .     ‚Äî      ,       : , -, UI.  Es decir     ; </li><li>        ,   :      ,      <code>catch</code> ; </li><li>       ‚Äì      .        ; </li><li>             ,       .      :       ,         , ,  . ,   -    :        ,   ‚Äî  ,  , ,        ; </li><li>         </li><li>       ,      :     ; </li><li>           ,  Mixed Mode c ErrorCode. </li></ul><br><h3 id="po-istochniku-oshibki">    </h3><br><p>             . ,     ,     : </p><br><ul><li>  unsafe ,    .      :         ,      (,   )     ; </li><li>     ,   ,      , ..       .        ,             ,     .  ,       ,       .            ‚Äî        ‚Äî  <code>InnerExcepton</code> .            ‚Äî          ; </li><li><font style="vertical-align: inherit;"><font style="vertical-align: inherit;">Nuestro propio c√≥digo que se ingres√≥ aleatoriamente en un estado no consistente. </font><font style="vertical-align: inherit;">Analizar texto es un buen ejemplo. </font><font style="vertical-align: inherit;">No hay dependencias externas, no hay retirada </font></font><code>unsafe</code><font style="vertical-align: inherit;"><font style="vertical-align: inherit;">, pero hay un error de an√°lisis.</font></font></li></ul><br><blockquote><h3>  Enlace a todo el libro </h3><br><ul><li><img src="https://habrastorage.org/webt/3q/6g/qa/3q6gqaz40qx-jzscjf3jbxatxhg.png">  Libro CLR: <a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=es&amp;u=">GitHub</a> </li><li><img src="https://habrastorage.org/webt/eo/6g/eo/eo6geog0tg5ernqmv2lcmufefta.png">  Libros de la versi√≥n 0.5.0, PDF: <a href="">versi√≥n de GitHub</a> </li></ul><br></blockquote><p> <a href="https://translate.googleusercontent.com/translate_c?depth=1&amp;rurl=translate.google.com&amp;sl=ru&amp;sp=nmt4&amp;tl=es&amp;u="><img src="https://habrastorage.org/getpro/habr/post_images/97f/1d3/cf0/97f1d3cf0e2a6bf007066eb60a789c31.png"></a> </p></div></div><p>Source: <a rel="nofollow" href="https://habr.com/ru/post/es419927/">https://habr.com/ru/post/es419927/</a></p>
<section class="more-articles-navigation-panel js-more-articles-navigation-panel">
<h4>More articles:</h4>
<nav class="list-of-articles-container js-list-of-articles-container"><ul class="list-of-pages js-list-of-pages">
<li><a href="../es419917/index.html">Redes neuronales: implementaci√≥n de la tarea sobre hongos en Tensor Flow y Python</a></li>
<li><a href="../es419919/index.html">Control de versiones dentro de SQL Server</a></li>
<li><a href="../es419921/index.html">C√≥mo soltar 10 millones de paquetes por segundo</a></li>
<li><a href="../es419923/index.html">Mi trabajo temporal, relojes de placa base</a></li>
<li><a href="../es419925/index.html">Control de versiones de archivos individuales usando GitHub Gist</a></li>
<li><a href="../es419929/index.html">[DotNetBook] Eventos de excepci√≥n y c√≥mo obtener StackOverflow y ExecutionEngineException desde cero</a></li>
<li><a href="../es419931/index.html">[DotNetBook] Tiempo para historias entretenidas: situaciones excepcionalmente excepcionales</a></li>
<li><a href="../es419933/index.html">C√≥mo hacer b√∫squedas de usuarios en Github usando Angular</a></li>
<li><a href="../es419935/index.html">Exactamente una vez NO es exactamente lo mismo: an√°lisis de art√≠culos</a></li>
<li><a href="../es419939/index.html">C√≥mo hice que la navegaci√≥n en React Native no sea tan terrible</a></li>
</ul></nav>
</section><br />
<a href="../../allArticles.html"><strong>All Articles</strong></a>
<script src="../../js/main.js"></script>

<!-- Yandex.Metrika counter -->
<script type="text/javascript" >
  (function (d, w, c) {
      (w[c] = w[c] || []).push(function() {
          try {
              w.yaCounter57283870 = new Ya.Metrika({
                  id:57283870,
                  clickmap:true,
                  trackLinks:true,
                  accurateTrackBounce:true,
                  webvisor:true
              });
          } catch(e) { }
      });

      var n = d.getElementsByTagName("script")[0],
          s = d.createElement("script"),
          f = function () { n.parentNode.insertBefore(s, n); };
      s.type = "text/javascript";
      s.async = true;
      s.src = "https://mc.yandex.ru/metrika/watch.js";

      if (w.opera == "[object Opera]") {
          d.addEventListener("DOMContentLoaded", f, false);
      } else { f(); }
  })(document, window, "yandex_metrika_callbacks");
</script>
<noscript><div><img src="https://mc.yandex.ru/watch/57283870" style="position:absolute; left:-9999px;" alt="" /></div></noscript>

<!-- Google Analytics -->
  <script>
    window.ga = function () { ga.q.push(arguments) }; ga.q = []; ga.l = +new Date;
    ga('create', 'UA-134228602-6', 'auto'); ga('send', 'pageview')
  </script>
  <script src="https://www.google-analytics.com/analytics.js" async defer></script>

</section>

<footer class="page-footer">
  <div class="page-footer-legal-info-container page-footer-element">
    <p>
      Weekly-Geekly ES | <span class="page-footer-legal-info-year js-page-footer-legal-info-year">2019</span>
    </p>
  </div>
  <div class="page-footer-counters-container page-footer-element">
    <a class="page-footer-counter-clustrmap" href='#'  title='Visit tracker'><img src='https://clustrmaps.com/map_v2.png?cl=698e5a&w=271&t=t&d=9uU9J9pq8z7k8xEBHYSfs6DenIBAHs3vLIHcPIJW9d0&co=3a3a3a&ct=ffffff'/></a>
  </div>
</footer>
  
</body>

</html>